tests/test_precondition.py

```diff
@@ -0,0 +1,88 @@
+# we are generating noisy test data using sine and cosine functions with multiple frequencies
+import unittest
+from aopy.visualization import *
+import matplotlib.pyplot as plt
+from aopy import precondition
+import time
+
+test_dir = os.path.dirname(__file__)
+write_dir = os.path.join(test_dir, 'tmp')
+if not os.path.exists(write_dir):
+    os.mkdir(write_dir)
+
+
+class FilterTests(unittest.TestCase):
+
+    def __init_subclass__(self, **kwargs):
+        self.T = 0.05
+        self.fs = 25000
+        self.freq = [600, 312, 2000]
+        self.a = 0.02
+        self.f0 = self.freq[0]
+        # testing generate test_signal
+        _x, _t = precondition.generate_test_signal(self.T, self.fs, self.freq, self.a)
+
+    def setUp(self):
+        self.T = 0.05
+        self.fs = 25000
+        self.freq = [600, 312, 2000]
+        self.a = 0.02
+        # testing generate test_signal
+        self.f0 = self.freq[0]
+        _x, _t = precondition.generate_test_signal(self.T, self.fs, self.freq, [self.a * 2, self.a*0.5, self.a*1.5, self.a*20 ])
+
+        self.x = _x
+        self.t = _t
+
+    def test_butterworth(self):
+        # Sample rate and desired cutoff frequencies (in Hz).
+        # fs = 25000.0
+        lowcut = 500.0
+        highcut = 1200.0
+        tic = time.perf_counter()
+        x_filter = precondition.butterworth_filter_data(self.x, lowcut, highcut, self.fs)
+        toc = time.perf_counter()
+        print(f" Butterworth filter executed in {toc - tic:0.4f} seconds")
+
+        fname = 'test_signal_filtered_Signal.png'
+        plot_filtered_signal(self.t, self.x, x_filter, lowcut, highcut)
+        plt.show()
+        savefig(write_dir, fname)
```

Comment:
something i should also add to my tests in visualization.py -- a comment here saying what to expect from the figure so reviewers know what to check for